contracts/Staking.sol

```diff
@@ -105,123 +122,168 @@ contract Staking is IStaking, System, IParamSubscriber, IApplication {
+      ++idx;
+    }
+    require(success, "rlp decode failed");
+
+    require(_checkPackHash(packBytes), "wrong pack hash");
+    iter = packBytes.toRLPItem().iterator();
     uint8 eventType = uint8(iter.next().toUint());
+    RLPDecode.Iterator memory paramIter;
+    if (iter.hasNext()) {
+      paramIter = iter.next().toBytes().toRLPItem().iterator();
+    } else {
+      revert("empty ack package");
+    }
     if (eventType == EVENT_DELEGATE) {
-      _handleDelegateAckPackage(iter);
+      _handleDelegateAckPackage(paramIter, status, errCode);
     } else if (eventType == EVENT_UNDELEGATE) {
-      _handleUndelegateAckPackage(iter);
+      _handleUndelegateAckPackage(paramIter, status, errCode);
     } else if (eventType == EVENT_REDELEGATE) {
-      _handleRedelegateAckPackage(iter);
+      _handleRedelegateAckPackage(paramIter, status, errCode);
     } else {
-      require(false, "unknown event type");
+      revert("unknown event type");
     }
-    return;
   }
 
   function handleFailAckPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override {
+    require(_checkPackHash(msgBytes), "wrong pack hash");
     RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
     uint8 eventType = uint8(iter.next().toUint());
+    RLPDecode.Iterator memory paramIter;
+    if (iter.hasNext()) {
+      paramIter = iter.next().toBytes().toRLPItem().iterator();
+    } else {
+      revert("empty fail ack package");
+    }
     if (eventType == EVENT_DELEGATE) {
-      _handleDelegateFailAckPackage(iter);
+      _handleDelegateFailAckPackage(paramIter);
     } else if (eventType == EVENT_UNDELEGATE) {
-      _handleUndelegateFailAckPackage(iter);
+      _handleUndelegateFailAckPackage(paramIter);
     } else if (eventType == EVENT_REDELEGATE) {
-      _handleRedelegateFailAckPackage(iter);
+      _handleRedelegateFailAckPackage(paramIter);
     } else {
-      require(false, "unknown event type");
+      revert("unknown event type");
     }
     return;
   }
 
   /***************************** External functions *****************************/
-  function delegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+  function delegate(address validator, uint256 amount) override external payable noReentrant tenDecimalPrecision(amount) initParams {
     require(amount >= minDelegation, "invalid delegate amount");
-    require(msg.value >= amount.add(oracleRelayerFee), "not enough msg value");
+    require(msg.value >= amount.add(relayerFee), "not enough msg value");
     require(payable(msg.sender).send(0), "invalid delegator"); // the msg sender must be payable
 
     uint256 convertedAmount = amount.div(TEN_DECIMALS); // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
-    uint256 _oracleRelayerFee = (msg.value).sub(amount);
+    uint256 _relayerFee = (msg.value).sub(amount);
+    uint256 oracleRelayerFee = _relayerFee>>1;
```

Comment:
`_relayerFee` can dive tenDecimal, but `_relayerFee>>1` may not, we need ensure `oracleRelayerFee` can divide tenDecimal.

Reply:
Done